src/Microsoft.OData.Core/MultipartMixed/ODataMultipartMixedBatchWriter.cs

```diff
@@ -0,0 +1,517 @@
+        }
+#endif
+
+        /// <summary>
+        /// Starts a new changeset - implementation of the actual functionality.
+        /// </summary>
+        protected override void WriteStartChangesetImplementation()
+        {
+            // write pending message data (headers, response line) for a previously unclosed message/request
+            this.WritePendingMessageData(true);
+
+            // important to do this first since it will set up the change set boundary.
+            this.SetState(BatchWriterState.ChangesetStarted);
+            Debug.Assert(this.changeSetBoundary != null, "this.changeSetBoundary != null");
+
+            // write the boundary string
+            ODataMultipartMixedBatchWriterUtils.WriteStartBoundary(this.rawOutputContext.TextWriter, this.batchBoundary, !this.batchStartBoundaryWritten);
+            this.batchStartBoundaryWritten = true;
+
+            // write the change set headers
+            ODataMultipartMixedBatchWriterUtils.WriteChangeSetPreamble(this.rawOutputContext.TextWriter, this.changeSetBoundary);
+            this.changesetStartBoundaryWritten = false;
+        }
+
+        /// <summary>
+        /// Creates an <see cref="ODataBatchOperationRequestMessage"/> for writing an operation of a batch request - implementation of the actual functionality.
+        /// </summary>
+        /// <param name="method">The Http method to be used for this request operation.</param>
+        /// <param name="uri">The Uri to be used for this request operation.</param>
+        /// <param name="contentId">The Content-ID value to write in ChangeSet head.</param>
+        /// <param name="payloadUriOption">The format of operation Request-URI, which could be AbsoluteUri, AbsoluteResourcePathAndHost, or RelativeResourcePath.</param>
+        /// <returns>The message that can be used to write the request operation.</returns>
+        protected override ODataBatchOperationRequestMessage CreateOperationRequestMessageImplementation(string method, Uri uri, string contentId, BatchPayloadUriOption payloadUriOption)
+        {
+            // write pending message data (headers, response line) for a previously unclosed message/request
+            this.WritePendingMessageData(true);
+
+            // Add a potential Content-ID header to the URL resolver so that it will be available
+            // to subsequent operations.
+            // Note that what we add here is the Content-ID header of the previous operation (if any).
+            // This also means that the Content-ID of the last operation in a changeset will never get
+            // added to the cache which is fine since we cannot reference it anywhere.
+            if (this.CurrentOperationContentId != null)
+            {
+                AddToPayloadUriConverter(this.CurrentOperationContentId);
+            }
+
+            this.InterceptException(() => uri = CreateOperationRequestUriWrapper(uri, this.rawOutputContext.MessageWriterSettings.BaseUri));
+

```

Comment:
This logic should be in the base class; it's not format-specific.  That should allow you to get rid of the (new) protected CreateOperationRequiestUriWrapper.

Reply:
Because we want to make this.payloadUriConverter private in base class, this wrapper method is needed to get access , as in "ODataBatchUtils.CreateOperationRequestUri(uri, baseUri, this.payloadUriConverter);"
Same for the other method.

---
In reply to: [143607061](https://github.com/OData/odata.net/pull/960#discussion_r143607061) [](ancestors = 143607061)

Reply:
You're missing the bigger picture.  There is nothing format-specific about creating the Uri; this should be done in the base class, with the wrapped Uri passed into CreateOperationRequestMesageImplementation.

---
In reply to: [143611607](https://github.com/OData/odata.net/pull/960#discussion_r143611607) [](ancestors = 143611607,143607061)

Reply:
Discussed offline. Move the payloadConverter & uri related logic to base class, but BuildOperationRequestMessage is still used due to access to OutputStream only available here.

---
In reply to: [144114063](https://github.com/OData/odata.net/pull/960#discussion_r144114063) [](ancestors = 144114063,143611607,143607061)